overlord/snapstate/handlers_components_link_test.go

```diff
@@ -0,0 +1,232 @@
+	s.AddCleanup(snapstate.MockTimeNow(func() time.Time {
+		return taskRunTime
+	}))
+}
+
+func (s *linkCompSnapSuite) TestDoLinkComponent(c *C) {
+	const snapName = "mysnap"
+	const compName = "mycomp"
+	snapRev := snap.R(1)
+	compRev := snap.R(7)
+	si := createTestSnapInfoForComponent(c, snapName, snapRev, compName)
+	ssu := createTestSnapSetup(si, snapstate.Flags{})
+
+	s.state.Lock()
+
+	// state does not contain the component
+	setStateWithOneSnap(c, s.state, snapName, snapRev)
+
+	t := s.state.NewTask("link-component", "task desc")
+	cref := naming.NewComponentRef(snapName, compName)
+	csi := snap.NewComponentSideInfo(cref, compRev)
+	t.Set("component-setup", snapstate.NewComponentSetup(csi, ""))
+	t.Set("snap-setup", ssu)
+	chg := s.state.NewChange("test change", "change desc")
+	chg.AddTask(t)
+
+	s.state.Unlock()
+
+	s.se.Ensure()
+	s.se.Wait()
+
+	s.state.Lock()
+
+	c.Check(chg.Err(), IsNil)
+	// undo information has been stored
+	var storedCsi snap.ComponentSideInfo
+	t.Get("linked-component", &storedCsi)
+	c.Assert(&storedCsi, DeepEquals, csi)
+	// state is modified as expected
+	var snapst snapstate.SnapState
+	c.Assert(snapstate.Get(s.state, snapName, &snapst), IsNil)
+	stateCsi := snapst.CurrentComponentSideInfo(cref)
+	c.Assert(stateCsi, DeepEquals, csi)
+	c.Assert(snapst.LastCompRefreshTime[csi.Component.ComponentName], Equals, taskRunTime)
+
+	s.state.Unlock()
+}
+
+func (s *linkCompSnapSuite) TestDoLinkThenUndoLinkComponent(c *C) {
+	const snapName = "mysnap"
+	const compName = "mycomp"
+	snapRev := snap.R(1)
+	compRev := snap.R(7)
+	si := createTestSnapInfoForComponent(c, snapName, snapRev, compName)
+	ssu := createTestSnapSetup(si, snapstate.Flags{})
+
+	s.state.Lock()
+
+	// state does not contain the component
+	setStateWithOneSnap(c, s.state, snapName, snapRev)
+
+	t := s.state.NewTask("link-component", "task desc")
+	cref := naming.NewComponentRef(snapName, compName)
+	csi := snap.NewComponentSideInfo(cref, compRev)
+	t.Set("component-setup", snapstate.NewComponentSetup(csi, ""))
+	t.Set("snap-setup", ssu)
+	chg := s.state.NewChange("test change", "change desc")
+	chg.AddTask(t)
+
+	terr := s.state.NewTask("error-trigger", "provoking undo link")
+	terr.WaitFor(t)
+	chg.AddTask(terr)
+
+	s.state.Unlock()
+
+	for i := 0; i < 3; i++ {
+		s.se.Ensure()
+		s.se.Wait()
+	}
+
+	s.state.Lock()
+
+	c.Check(chg.Err().Error(), Equals, "cannot perform the following tasks:\n"+
+		"- provoking undo link (error out)")
+	// undo information was stored
+	var storedCsi snap.ComponentSideInfo
+	t.Get("linked-component", &storedCsi)
+	c.Assert(&storedCsi, DeepEquals, csi)
+	// the component is not in the state
+	var snapst snapstate.SnapState
+	c.Assert(snapstate.Get(s.state, snapName, &snapst), IsNil)
+	c.Assert(snapst.CurrentComponentSideInfo(cref), IsNil)
+	c.Assert(snapst.LastCompRefreshTime[csi.Component.ComponentName], Equals, taskRunTime)
+
+	s.state.Unlock()
+}
+
+func (s *linkCompSnapSuite) TestDoUnlinkComponent(c *C) {
```

Comment:
shouldn't this names mention Current? as we will likely also have unlink-component itself?

Reply:
Good point, changed